packages/editor/src/components/page-attributes/parent.js

```diff
@@ -1,44 +1,195 @@
 /**
  * External dependencies
  */
-import { get } from 'lodash';
+import { get, debounce } from 'lodash';
+import Autocomplete from 'accessible-autocomplete/react';
 
 /**
  * WordPress dependencies
  */
-import { __ } from '@wordpress/i18n';
+import { sprintf, __, _n } from '@wordpress/i18n';
 import { TreeSelect } from '@wordpress/components';
 import { compose } from '@wordpress/compose';
 import { withSelect, withDispatch } from '@wordpress/data';
+import { Component } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
 
 /**
  * Internal dependencies
  */
 import { buildTermsTree } from '../../utils/terms';
 
-export function PageAttributesParent( { parent, postType, items, onUpdateParent } ) {
-	const isHierarchical = get( postType, [ 'hierarchical' ], false );
-	const parentPageLabel = get( postType, [ 'labels', 'parent_item_colon' ] );
-	const pageItems = items || [];
-	if ( ! isHierarchical || ! parentPageLabel || ! pageItems.length ) {
-		return null;
+export class PageAttributesParent extends Component {
+	constructor() {
+		super( ...arguments );
+		this.searchCache = [];
+		this.getCurrentParentFromAPI = this.getCurrentParentFromAPI.bind( this );
+		this.handleSelection = this.handleSelection.bind( this );
+		this.suggestPage = this.suggestPage.bind( this );
+
+		this.requestResults = debounce( ( query, populateResults ) => {
+			const payload = '?search=' + encodeURIComponent( query );
+			apiFetch( { path: `/wp/v2/pages${ payload }` } ).then( ( results ) => {
+				populateResults( this.resolveResults( results ) );
+				this.searchCache[ query ] = results;
+			} );
+		}, 300 );
+		this.state = {
+			parentPost: false,
+		};
+	}
+
+	/**
+	 * Retrieve the parent page by id.
+	 *
+	 * @param {number} parentId The id of the parent to fetch.
+	 */
+	async getCurrentParentFromAPI( parentId ) {
+		if ( ! parentId ) {
+			return '';
+		}
+		const parentPost = await apiFetch( { path: `/wp/v2/pages/${ parentId }` } );
+		this.setState( {
+			parentPost,
+		} );
+	}
+
+	/**
+	 * Resolve the results for display.
+	 *
+	 * @param {Array} results The array of pages that matched the search.
+	 *
+	 * @return {Array} an array of strings ready for displaying.
+	 */
+	resolveResults( results ) {
+		return results.map( ( item ) => item.title.rendered ? `${ item.title.rendered } (#${ item.id })` : `${ __( 'no title' ) } (#${ item.id })` );
 	}
 
-	const pagesTree = buildTermsTree( pageItems.map( ( item ) => ( {
-		id: item.id,
-		parent: item.parent,
-		name: item.title.raw ? item.title.raw : `#${ item.id } (${ __( 'no title' ) })`,
-	} ) ) );
-	return (
-		<TreeSelect
-			className="editor-page-attributes__parent"
-			label={ parentPageLabel }
-			noOptionLabel={ `(${ __( 'no parent' ) })` }
-			tree={ pagesTree }
-			selectedId={ parent }
-			onChange={ onUpdateParent }
-		/>
-	);
+	handleSelection( selection ) {
```

Comment:
If I enter an empty string in the input, `selection` here is undefined, so it's effectively impossible to unset a parent page once it's been set. We should be able to handle an empty string as a `no parent` option.

Reply:
Thanks for the feedback @tellthemachines - my intention is definately to combine these PRs by creating our own accessible autocomplete component that wraps/abstracts the underlying external dependency. Part of doing this second PR was finding the common implementation points we need to implement. I welcome any help in that effort! :)

It's worth noting for context that the team has already implemented and reviewed accessible autocomplete components from several popular sources, all of which aim for/achieve some degree of accessibility. Of everything we reviewed, this one rose to the top and continues to receive active sustained development.

In terms of `Preact` as a requirement - that sounds like a bug or build issue - we will be sure to remove that as a dependency, the project does support Preact but does not require it.